WIP: Consider new SMS functionality as a PR for 2.32

JasperTimm opened the pull request:
NOTE: This is not intended to be a finished PR, it's simply a work in progress for people to get an idea of what the new SMS functionality looks like.

The below PR includes all new development completed so far on the new SMS functionality for DHIS2, which includes the new listeners for all new submissions types as well as the inclusion of the separate SMS compression library which can be seen here: [SMS compression](https://github.com/eHealthAfrica/dhis2-sms-compression)

For those that are reviewing the code, as well as any obvious problems you find, it would be good to focus on the existing error handling, especially those lines with comments as TODOs to confirm that the way errors are being handled is as intended.

zewolfe committed:
Add: WIP commit

zewolfe committed:
Add: Initial EnrollmentSMSListener

Commit:
feat: add NewSMSListener for sms decompression

Commit:
Merge pull request #1 from eHealthAfrica/feat/jasper/new-base-sms-listener

feat: add NewSMSListener for sms decompression

Commit:
feat: add TrackerEventSMSListener

zewolfe committed:
Merge branch 'eha-sms' into wip/george

Commit:
temp: push build file to eHA S3

Commit:
fix: ensure war file is made public

Commit:
fix: try to separate acl change

Commit:
fix: logging and equality check in TrackerEventSMSListener

Commit:
Merge pull request #2 from eHealthAfrica/tracker-event-listener

feat: add TrackerEventSMSListener

Commit:
fix: ensure the TrackerEventSMSListener is listed in beans.xml

Commit:
Merge pull request #3 from eHealthAfrica/tracker-event-listener

fix: ensure the TrackerEventSMSListener is listed in beans.xml

Commit:
fix: fixes for TrackerEventSMSListener

Commit:
Merge pull request #4 from eHealthAfrica/tracker-event-listener

Fixes for TrackerEventSMSListener

zewolfe committed:
Merge branch 'eha-sms' into wip/george

zewolfe committed:
Add: Tracker Enrollments via SMS

zewolfe committed:
Update: Fix beans

zewolfe committed:
Update: Clean up

zewolfe committed:
Update: Remove Fake metadata

zewolfe committed:
Chore: Clean up

Commit:
Merge pull request #5 from eHealthAfrica/wip/george

Add: Tracker Enrolment via SMS

zewolfe committed:
Add: UID suuport for tracker enronlments

Commit:
fix: use the existing list method for enrolment

Commit:
Merge pull request #6 from eHealthAfrica/update/george/enrolments-sms

Add: UID suuport for tracker enrolments

Commit:
Pull in 2.32 and resolve conflicts

Commit:
Fix for enrolment listener after 2.32 merge

Commit:
Merge branch '2.32' of github.com:dhis2/dhis2-core into jasper-2.32-merge

Commit:
Merge pull request #8 from eHealthAfrica/jasper-2.32-merge

Merge in 2.32 branch from origin

vietnguyen committed:
Fix bug missing trackedEntityAttribute when importing trackedEntityTypeAttribute

Commit:
Update sms comp to 0.0.12

Commit:
Trigger dhis2 deploy to stage

Commit:
Feat: Add support for simple events, fix event UID in tracker events

Commit:
Merge pull request #9 from eHealthAfrica/simple-events

Feat: Add support for simple events, fix event UID in tracker events

Commit:
Fix: rename a few classes to differentiate from DHIS2 classes

Commit:
Merge pull request #10 from eHealthAfrica/rename-sms-classes

Fix: rename a few classes to differentiate from DHIS2 classes

Commit:
Feat: AgreggateDatasetSMSListener happy path works

Commit:
Merge pull request #11 from eHealthAfrica/aggregate-datasets

Feat: AgreggateDatasetSMSListener happy path works

Commit:
Feat: start of message id for SMS impl

Commit:
Feat: error handling with message id

Commit:
Merge pull request #12 from eHealthAfrica/message-id

Better error handling support with message IDs and error codes

Commit:
Feat: Delete Event Listener

Commit:
Merge pull request #13 from eHealthAfrica/delete-listener

Feat: Delete Event Listener

Commit:
Feat: Relationship Listener

Commit:
Merge pull request #14 from eHealthAfrica/relationship-listener

Feat: Relationship Listener

Commit:
Feat: better error handling in SMS listeners

Commit:
Feat: partial error handling

Commit:
Merge pull request #15 from eHealthAfrica/partial-errors

Partial errors

Commit:
Docs: using styling conf and copyright info

zubaira commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/AggregateDatasetSMSListener.java:
```diff
@@ -0,0 +1,191 @@
+package org.hisp.dhis.sms.listener;
+
```
The license text is missing

JasperTimm replied:
Yeah I had an outstanding PR with the licensing and styling updates, it's merged now. I also added the license to some existing files that didn't have it.

zubaira commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    public List<SMSMetadata.ID> getAllOrgUnitIds (Date lastSyncDate) {
+        List<OrganisationUnit> orgUnits = organisationUnitService.getAllOrganisationUnits();
```
these are heavy operations. Wouldn't it be good if we filter at SQL level?

JasperTimm replied:
Yeah, that's a good point, I suppose if we just need the IDs we shouldn't run a query to get all the metadata.

Could you point me to an example in the code where something like this has been done before? As in we would replace it with a hard coded SQL query?

zubaira commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    public List<SMSMetadata.ID> getAllDataElements (Date lastSyncDate) {
+        List<DataElement> dataElements = dataElementService.getAllDataElements();
```
same as above

zubaira commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java:
```diff
@@ -0,0 +1,370 @@
+		}
+	}
+
+	public SMSMetadata getMetadata( Date lastSyncDate ) {
```
can be private

JasperTimm replied:
Sure, I'll update these

zubaira commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java:
```diff
@@ -0,0 +1,370 @@
+		return meta;
+	}
+
+    public List<SMSMetadata.ID> getAllUserIds (Date lastSyncDate) {
```
same as above

zubaira commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java:
```diff
@@ -0,0 +1,370 @@
+	}
+	
+	private SMSSubmissionHeader getHeader( IncomingSms sms ) {
+		byte[] smsBytes = SmsUtils.getBytes(sms);
```
In case of any error in decoding SmsUtils.getBytes will return null

JasperTimm replied:
Yeah I can send an SMS error back in getHeader as well, the trick is that I won't have a message ID to tell the app which message failed. I'm leaning towards just sending back a message with a '?' for the message ID though so at least the app has something.

Commit:
Merge pull request #17 from eHealthAfrica/styling-change

Docs: using styling conf and copyright info

Commit:
chore: add licensing to some files

Commit:
style: change some public methods to private

Commit:
fix: improve error handling for null SMS header

Commit:
Merge pull request #18 from eHealthAfrica/review-updates

Review updates - error handling and style

Commit:
fix: remove imports of SMSResponse

Commit:
Merge pull request #19 from eHealthAfrica/review-updates

fix: remove imports of SMSResponse

Commit:
fix: change TEI to enrolment and event completion for SMS

Commit:
Merge pull request #20 from eHealthAfrica/update-for-sms-comp

fix: change TEI to enrolment and event completion for SMS

Commit:
fix: get new ver of SMS comp for event type fix

Commit:
refactor: move SMSResponse to SMS comp lib

Commit:
feat: update SMS listeners for comp lib 0.1.3

Commit:
Merge pull request #21 from eHealthAfrica/update-for-comp-0.1.3

feat: update SMS listeners for comp lib 0.1.3

Commit:
Merge branch '2.32' of github.com:dhis2/dhis2-core into eha-sms-2.32

Commit:
Merge pull request #22 from eHealthAfrica/eha-sms-2.32

Merge in latest 2.32 updates

Commit:
fix: move SMS comp lib dep

Commit:
test: adding the start of the newSMS tests

Commit:
fix: move checkUser and allow for throwables in SMSExcep

Commit:
Merge pull request #23 from eHealthAfrica/fix-exceptions

fix: move checkUser and allow for throwables in SMSExcep

Bekkalizer committed:
fix: DHIS2-7194 avoid parsing dates into daylight savings transition gap

vietnguyen committed:
Add back PDF datavalue import function for DataValueSetController

stian-sandvold committed:
Add Predictor percentileCont, stddevSamp, stddevPop

Issue:  DHIS2-6327
(cherry picked from commit bc1155f8a2adcb05c4b8c4eda54054f035781131)

vietnguyen committed:
Add sharing access validation for TEA when updating Program

vietnguyen committed:
Merge pull request #3488 from vietnguyen/DHIS2-7255-32

2.32 Add back PDF datavalue import function for DataValueSetController

vietnguyen committed:
Merge pull request #3491 from vietnguyen/DHIS2-6133-32

2.32 Fix bug missing TEA when importing TETA

ameenhere committed:
Fix redis performance issues
1. Upgrade spring session data redis, lettuce and spring data redis dependency
Remove notifiers in leader election job
3. Remove expireAfterAccess usage in UserSettingService and SystemSettingService

ameenhere committed:
Fix redis performance issues
1. Upgrade spring session data redis, lettuce and spring data redis dependency
Remove notifiers in leader election job
3. Remove expireAfterAccess usage in UserSettingService and SystemSettingService

vietnguyen committed:
Fix merge conflicts

vietnguyen committed:
Merge pull request #3496 from vietnguyen/DHIS2-6350-32

2.32 Add sharing access validation for TEA when updating Program

ameenhere committed:
Merge pull request #3498 from dhis2/DHIS2-6691-2.32

fix: redis performance issue

ameenhere committed:
fix: uninstalling app async process
1. Add app status DELETION_IN_PROGRESS
2. Make delete app method async
3. Render app fails if app is being deleted
4. Listing apps does not include app that is currently being deleted
5. Installing same app when the app is being deleted errors out

DavidCKen committed:
Backport of lastUpdatedDuration feature for Tracker API endpoints

- Add support for lastUpdatedDuration parameter to TEIs endpoint
- Fix a bug with using incorrect count method in validation
- Add support for lastUpdatedDuration parameter to Enrollments endpoint
- Add support for lastUpdatedDuration parameter to Events endpoint

DHIS2-6188

ameenhere committed:
Merge pull request #3509 from dhis2/DHIS2-6572-2.32

fix: uninstalling app async process (2.32)

ameenhere committed:
Add missing Async annotation

ameenhere committed:
Merge pull request #3514 from dhis2/DHIS2-6572-2.32

Add missing Async annotation

Commit:
fix: Handle non-POINT Geometry from ProgramInstance on table export

- DHIS2-7279
- Only populate analytics enrollment table's latitude/longitude if the
Program Instance geometry column is of type POINT

ameenhere committed:
fix: completing an enrollment and adding a new enrollment in the same payload fails

ameenhere committed:
Merge pull request #3523 from dhis2/DHIS2-7067-2.32

fix: completing an enrollment and adding a new enrollment in the same payload fails (2.32)

Commit:
fix: use setList in SMSResponse

vietnguyen committed:
Fix bug lastLogin reset after updating user

Commit:
Merge pull request #3521 from dhis2/2.32_DHIS2-7279_AnalyticsExportFailOnPolygon

[2.32] fix: Handle non-POINT Geometry from ProgramInstance on table export

Commit:
fix: working test for NewSMSListeners

Commit:
test: add remaining tests for new SMS listeners

Commit:
Merge pull request #24 from eHealthAfrica/new-sms-tests

Add tests for the new SMS listeners

vietnguyen committed:
Merge pull request #3529 from vietnguyen/DHIS2-7282-32

2.32 Fix bug lastLogin reset after updating user

DavidCKen committed:
Backport of lastUpdatedDuration feature for Tracker API endpoints

- Add support for lastUpdatedDuration parameter to TEIs endpoint
- Fix a bug with using incorrect count method in the validation
- Add support for lastUpdatedDuration parameter to Enrollments endpoint
- Add support for lastUpdatedDuration parameter to Events endpoint

DHIS2-6188

Commit:
fix: incorrect ids being stored for users in SMS metadata

DavidCKen committed:
Fix issue with DataElement identifiers not being translated to Code

Fix issue with DataElement identifiers not being translated to Code when idScheme=CODE is used.

DHIS2-7268

stian-sandvold committed:
Fix issue with DataElement identifiers not being translated to Code

Issue: DHIS2-7268

Commit:
Fix reporting rate calculation for daily data and months less than 30 days (#3501) (#3540)

* fix: reporting rate calculation for months with less then 30 days

- DHIS2-2110
- Fixed issue where a Daily dataset with 100% reporting rate would
be reported with a lower reporting rate %, for months with less then 30
days

* Method rename

(cherry picked from commit 0ca1a1fd7e564b29ba541f21bd5f55482dea0011)

stian-sandvold committed:
Fix FileResourceContent streaming closing prematurely

larshelge committed:
Use fluent pattern for EventSearchParams

stian-sandvold committed:
Fix FileResourceContent streaming closing prematurely (#3554)

* Fix FileResourceContent streaming closing prematurely

* minor style fix

Commit:
Fix logging message

Commit:
Remove unwanted text in code

larshelge committed:
Add param  to event GET endpoint

larshelge committed:
Set no cache/store header for settings and data store API

stian-sandvold committed:
Add method for refreshing status of jobConfiguration (#3572)

Issue: DHIS2-7237
(cherry picked from commit da70390c36cea27058b180ef8cf32766f6218742)

varl committed:
feat: support wildcards in cors domain matching (#3451) (#3562)

Fixes DHIS2-5592

Commit:
Add nested indicator support in expressions (#3580)

- DHIS2-6901
- Added support for expression supporting nested indicators (an indicator
                                referencing another indicator)
- Added detection for cyclic dependency between indicators

backport from: 82b263c89264c31fd8868d97ee1b65a016409be2

Commit:
fix: error get Enrollments through `/api/enrollments.json` endpoint (#3581) (#3583)

- DHIS2-7323
- Fixed by making Audit Service to run in its own TX context

(cherry picked from commit c118a4b013148ea11a0ce30d115cb0ccb5818401)

larshelge committed:
Set safety switch off for CSV metadata read

larshelge committed:
Merge branch '2.32' of github.com:dhis2/dhis2-core into 2.32

larshelge committed:
Use display property in event analytics headers [DHIS2-1790] (#3588)

Ensure the display property based on the system setting is used in event analytics responses for the header part.

Fixes bug DHIS2-1790.

(cherry picked from commit 53f2d3b50d4a8d56e7ad43fc726a860e11d157b3)

larshelge committed:
Fix crash with event analytics query due to column headers

larshelge committed:
Fix crash with event analytics query due to column headers (#3596)

vilkg committed:
2.32: removing stopped containers on jenkins (#3592)

* Removing stopped containers

* Dockerfile: Nesting RUN commands

vilkg committed:
Removing stopped containers

stian-sandvold committed:
Clean jenkins docker configuration

Commit:
fix: set Geometry field on ProgramStageInstance when updating event

(cherry picked from commit 79d9f8a821c2ceefeba2bd6e8462846ac3355861)

Commit:
Merge pull request #3604 from dhis2/2.32_DHIS2-7372

[2.32] fix: set Geometry field on ProgramStageInstance when updating event

vietnguyen committed:
Add translations for DataSet Section

vietnguyen committed:
Merge pull request #3609 from vietnguyen/DHIS2-6135-32

2.32 Add translations for DataSet Section

larshelge committed:
Extend period cache timeout (#3612)

Fix wrongly set period cache timeout.

(cherry picked from commit 1c37c9516f2d560301971f06df72d7fff76d0cde)

larshelge committed:
Merge branch '2.32' of github.com:dhis2/dhis2-core into 2.32

ameenhere committed:
fix: add reference in importsummary when deleting events and enrollments

ameenhere committed:
Merge branch '2.32' into DHIS2-6914-2.32

larshelge committed:
fix: Add reference in importsummary when deleting events and enrollments (2.32)

Merge pull request #3617 from dhis2/DHIS2-6914-2.32

Commit:
Merge branch '2.32' of github.com:dhis2/dhis2-core into eha-sms

JasperTimm commented:
Closing this and opening up a new PR
